src/test/app/PermissionedDomains_test.cpp

```diff
@@ -0,0 +1,504 @@
+            jss::Issuer);
+        BEAST_EXPECT(
+            exceptionExpected(env, txJsonMutable).starts_with("invalidParams"));
+
+        // Make 2 identical credentials. The duplicate should be silently
+        // removed.
+        {
+            pdomain::Credentials const credentialsDup{
+                {alice7, "credential6"},
+                {alice2, "credential1"},
+                {alice3, "credential2"},
+                {alice2, "credential1"},
+                {alice5, "credential4"},
+            };
+
+            std::unordered_map<std::string, Account> pubKey2Acc;
+            for (auto const& c : credentialsDup)
+                pubKey2Acc.emplace(c.issuer.human(), c.issuer);
+
+            auto const sorted = pdomain::sortCredentials(credentialsDup);
+            BEAST_EXPECT(sorted.size() == 4);
+            env(pdomain::setTx(account, credentialsDup, domain),
+                ter(temMALFORMED));
+
+            env.close();
+            env(pdomain::setTx(account, sorted, domain));
+
+            uint256 d;
+            if (domain)
+                d = *domain;
+            else
+                d = pdomain::getNewDomain(env.meta());
+            env.close();
+            auto objects = pdomain::getObjects(account, env);
+            auto const fromObject =
+                pdomain::credentialsFromJson(objects[d], pubKey2Acc);
+            auto const sortedCreds = pdomain::sortCredentials(credentialsDup);
+            BEAST_EXPECT(fromObject == sortedCreds);
+        }
+
+        // Have equal issuers but different credentials and make sure they
+        // sort correctly.
+        {
+            pdomain::Credentials const credentialsSame{
+                {alice2, "credential3"},
+                {alice3, "credential2"},
+                {alice2, "credential9"},
+                {alice5, "credential4"},
+                {alice2, "credential6"},
+            };
+            std::unordered_map<std::string, Account> pubKey2Acc;
+            for (auto const& c : credentialsSame)
+                pubKey2Acc.emplace(c.issuer.human(), c.issuer);
+
+            BEAST_EXPECT(
+                credentialsSame != pdomain::sortCredentials(credentialsSame));
+            env(pdomain::setTx(account, credentialsSame, domain));
+
+            uint256 d;
+            if (domain)
+                d = *domain;
+            else
+                d = pdomain::getNewDomain(env.meta());
+            env.close();
+            auto objects = pdomain::getObjects(account, env);
+            auto const fromObject =
+                pdomain::credentialsFromJson(objects[d], pubKey2Acc);
+            auto const sortedCreds = pdomain::sortCredentials(credentialsSame);
+            BEAST_EXPECT(fromObject == sortedCreds);
+        }
+    }
+
+    // Test PermissionedDomainSet
+    void
+    testSet()
+    {
+        testcase("Set");
+        Env env(*this, withFeature_);
+        env.set_parse_failure_expected(true);
```

Comment:
Why does this expect a parse failure?

Reply:
When some field missed in tx, it throw an error like 
 ```
template_mismatch: Field 'Issuer' is required but missing. parse failed:
```
and put the whole transaction to the output.  This flag doesn't prevent exception throwing, it just silent unnecessary long and annoying output.  Tests are designed to output 1 row - name of the test and status Failed/Passed.  Such additional output makes you think that  something is wrong with the test, when it is not.